### PR TITLE
const-eval: disallow unwinding across functions that `!fn_can_unwind()`

### PR DESCRIPTION
**Pull request by ghost**
Following https://github.com/rust-lang/miri/pull/1776#discussion_r633074343, so r? @RalfJung

This PR turns `unwind` in `StackPopCleanup::Goto` into a new enum `StackPopUnwind`, with a `NotAllowed` variant to indicate that unwinding is not allowed. This variant is chosen based on `rustc_middle::ty::layout::fn_can_unwind()` in `eval_fn_call()` when pushing the frame. A check is added in `unwind_to_block()` to report UB if unwinding happens across a `StackPopUnwind::NotAllowed` frame.

Tested with Miri `HEAD` with [minor changes](https://github.com/rust-lang/miri/compare/HEAD..9cf3c7f0d86325a586fbcbf2acdc9232b861f1d8) and the rust-lang/miri#1776 branch with [these changes](https://github.com/rust-lang/miri/compare/d866c1c52f48bf562720383455b75c257bb1ad92..626638fbfe2fff34648dda29a34d59db498a6e52).

### CODE REVIEW

**Comment by RalfJung** [comment]
> Tested with Miri HEAD with minor changes 

Could you please prepare a Miri PR with just those changes? Then we can fix the toolstate ASAP without having to complete the other, much larger, PR.

File: compiler/rustc_middle/src/ty/layout.rs [hunk 1]
```diff
@@ -2641,6 +2641,43 @@ fn fn_can_unwind(
...
+
+        // These API constants ought to be more specific...
+        Cdecl => Conv::C,
+    }
+}
+
+pub fn fn_ptr_codegen_fn_attr_flags() -> CodegenFnAttrFlags {
+    // Assume that fn pointers may always unwind
+    CodegenFnAttrFlags::UNWIND
+}
```
Comment (RalfJung) [inline compiler/rustc_middle/src/ty/layout.rs:2679]: Why is this a function (and not, e.g., a `const`)?

File: compiler/rustc_mir/src/interpret/eval_context.rs [hunk 2]
```diff
@@ -134,14 +134,23 @@ pub struct FrameInfo<'tcx> {
     pub lint_root: Option<hir::HirId>,
 }
 
+/// Unwind information.
+#[derive(Clone, Copy, Eq, PartialEq, Debug, HashStable)]
+pub enum StackPopUnwind {
+    /// The cleanup block.
+    Cleanup(Option<mir::BasicBlock>),
```
Comment (RalfJung) [inline compiler/rustc_mir/src/interpret/eval_context.rs:141]: Please explain the situation where the block is `None`.

Actually, it's probably better to "bubble up" the `Option` like this:
```rust
Cleanup(mir::BasicBlock),
Skip,
NotAllowed,
```

File: compiler/rustc_mir/src/interpret/eval_context.rs [hunk 3]
```diff
@@ -807,9 +816,29 @@ impl<'mir, 'tcx: 'mir, M: Machine<'mir, 'tcx>> InterpCx<'mir, 'tcx, M> {
         // In that case, we return early. We also avoid validation in that case,
         // because this is CTFE and the final value will be thoroughly validated anyway.
         let (cleanup, next_block) = match frame.return_to_block {
-            StackPopCleanup::Goto { ret, unwind } => {
-                (true, Some(if unwinding { unwind } else { ret }))
-            }
+            StackPopCleanup::Goto { ret, unwind } => (
+                true,
+                Some(if unwinding {
```
Comment (RalfJung) [inline compiler/rustc_mir/src/interpret/eval_context.rs:821]: Please move this `if` up a bit, this is nesting control flow a bit too deeply. I think putting it in the match would work, like
```rust
match (frame.return_to_block, unwinding) {
  (StackPopCleanup::Goto { ret, _ }, false) => ...
  (StackPopCleanup::Goto { _, unwind }, true) => ...
  (StackPopCleanup::None { cleanup, .. }, _) => ...
}
```

File: compiler/rustc_mir/src/interpret/eval_context.rs [hunk 4]
```diff
@@ -807,9 +816,29 @@ impl<'mir, 'tcx: 'mir, M: Machine<'mir, 'tcx>> InterpCx<'mir, 'tcx, M> {
...
+                    let def_id = frame.body.source.def_id();
+                    match unwind {
+                        StackPopUnwind::Cleanup(unwind)
+                            // `fn_sig()` can't be used on closures, but closures always have
+                            // "rust-call" ABI, which always allows unwinding anyway.
+                            if self.tcx.is_closure(def_id) || self.fn_can_unwind(
+                                self.tcx.codegen_fn_attrs(def_id).flags,
+                                self.tcx.fn_sig(def_id).abi(),
+                            ) =>
```
Comment (RalfJung) [inline compiler/rustc_mir/src/interpret/eval_context.rs:830]: This is strange. We already have an `unwind` block here, why do we check something again?
The only UB rising in `pop_stack_frame` should be when there's `StackPopUnwind::NotAllowed`.

File: compiler/rustc_mir/src/interpret/terminator.rs [hunk 5]
```diff
@@ -12,9 +13,19 @@ use rustc_target::spec::abi::Abi;
 
 use super::{
     FnVal, ImmTy, InterpCx, InterpResult, MPlaceTy, Machine, OpTy, PlaceTy, StackPopCleanup,
+    StackPopUnwind,
 };
 
 impl<'mir, 'tcx: 'mir, M: Machine<'mir, 'tcx>> InterpCx<'mir, 'tcx, M> {
+    pub(super) fn fn_can_unwind(&self, attrs: CodegenFnAttrFlags, abi: Abi) -> bool {
```
Comment (RalfJung) [inline compiler/rustc_mir/src/interpret/terminator.rs:20]: This function not being private is suspicious. Why is that?

File: compiler/rustc_mir/src/interpret/terminator.rs [hunk 6]
```diff
@@ -220,6 +236,7 @@ impl<'mir, 'tcx: 'mir, M: Machine<'mir, 'tcx>> InterpCx<'mir, 'tcx, M> {
         args: &[OpTy<'tcx, M::PointerTag>],
         ret: Option<(&PlaceTy<'tcx, M::PointerTag>, mir::BasicBlock)>,
         unwind: Option<mir::BasicBlock>,
+        can_unwind: bool,
```
Comment (RalfJung) [inline compiler/rustc_mir/src/interpret/terminator.rs:239]: `unwind`+`can_unwind` combine exactly to a `StackPopUnwind`, why not pass that?

File: compiler/rustc_middle/src/ty/layout.rs (hunk 1 above)
Comment (hyd-dev) [reply compiler/rustc_middle/src/ty/layout.rs:2679]: The `|` operator for `CodegenFnAttrFlags` does not work in `const`, so if people start to add more flags here, it will break.

File: compiler/rustc_mir/src/interpret/eval_context.rs (hunk 2 above)
Comment (hyd-dev) [reply compiler/rustc_mir/src/interpret/eval_context.rs:141]: > Please explain the situation where the block is `None`.

I did not found the explanation before, so I didn't write it here.
I guess I can write "no cleanup needs to be done".

> Actually, it's probably better to "bubble up" the `Option`

`Option<mir::BasicBlock>` is used in some other places, so I chose `Cleanup(Option<_>)` for simplicity, but I can change it.

File: compiler/rustc_mir/src/interpret/eval_context.rs (hunk 4 above)
Comment (hyd-dev) [reply compiler/rustc_mir/src/interpret/eval_context.rs:830]: Right. The `if` should probably be moved to `eval_fn_call()`, after `M::find_mir_or_eval_fn()` (it's checking `source`). Not sure why I have put it here...

File: compiler/rustc_mir/src/interpret/terminator.rs (hunk 5 above)
Comment (hyd-dev) [reply compiler/rustc_mir/src/interpret/terminator.rs:20]: https://github.com/rust-lang/rust/pull/85546#discussion_r637166933 :facepalm:

File: compiler/rustc_mir/src/interpret/eval_context.rs [hunk 7]
```diff
@@ -134,14 +134,23 @@ pub struct FrameInfo<'tcx> {
...
+/// Unwind information.
+#[derive(Clone, Copy, Eq, PartialEq, Debug, HashStable)]
+pub enum StackPopUnwind {
+    /// The cleanup block.
+    Cleanup(Option<mir::BasicBlock>),
+    /// Unwinding is not allowed (UB).
+    NotAllowed,
+}
+
+#[derive(Clone, Copy, Eq, PartialEq, Debug, HashStable)] // Miri debug-prints these
```
Comment (hyd-dev) [inline compiler/rustc_mir/src/interpret/eval_context.rs:146]: I have to add `Copy` here, otherwise I will get this error (for https://github.com/rust-lang/rust/pull/85546#discussion_r637156732):
```
error[E0382]: use of partially moved value: `frame`
   --> compiler/rustc_mir/src/interpret/eval_context.rs:845:37
    |
818 |         let (cleanup, next_block) = match (frame.return_to_block, unwinding) {
    |                                            --------------------- value partially moved here
...
845 |         if M::after_stack_pop(self, frame, unwinding)? == StackPopJump::NoJump {
    |                                     ^^^^^ value used here after partial move
    |
    = note: partial move occurs because `frame.return_to_block` has type `StackPopCleanup`, which does not implement the `Copy` trait
```
(Or, maybe I have made a mistake when writing the `match`...)

File: compiler/rustc_middle/src/ty/layout.rs (hunk 1 above)
Comment (RalfJung) [reply compiler/rustc_middle/src/ty/layout.rs:2679]: Hm, I feel a `const fn` replacement for that should be possible... but, sure, that's a reasonable answer.

File: compiler/rustc_mir/src/interpret/eval_context.rs (hunk 2 above)
Comment (RalfJung) [reply compiler/rustc_mir/src/interpret/eval_context.rs:141]: > I did not found the explanation before, so I didn't write it here.

That's fair. :) But I think we should use this opportunity to fix this.

>  I guess I can write "no cleanup needs to be done".

Yes, that's what this means: on unwind, just pop this stack frame without running any cleanup code.

> Option<mir::BasicBlock> is used in some other places, so I chose Cleanup(Option<_>) for simplicity, but I can change it.

Yeah and it's rather hacky, since for *return* blocks `None` means "returning is UB" but for *unwind* blocks `None` means "no cleanup needed". I think we should clean this up, since now "unwinding is UB" also becomes a possibility so things just become too confusing.

File: compiler/rustc_mir/src/interpret/eval_context.rs (hunk 7 above)
Comment (RalfJung) [reply compiler/rustc_mir/src/interpret/eval_context.rs:146]: Making this type `Copy` seems totally fine for me.

File: compiler/rustc_mir/src/interpret/eval_context.rs [hunk 8]
```diff
@@ -815,31 +815,18 @@ impl<'mir, 'tcx: 'mir, M: Machine<'mir, 'tcx>> InterpCx<'mir, 'tcx, M> {
...
-                        {
-                            unwind
-                        }
-                        _ => {
-                            throw_ub_format!("unwind past a frame that does not allow unwinding")
-                        }
+                Some(match unwind {
+                    StackPopUnwind::Cleanup(unwind) => unwind,
+                    StackPopUnwind::NotAllowed => {
+                        throw_ub_format!("unwind past a frame that does not allow unwinding")
```
Comment (RalfJung) [inline compiler/rustc_mir/src/interpret/eval_context.rs:825]:
```suggestion
                        throw_ub_format!("unwinding past a frame that does not allow unwinding")
```
This is more consistent with some of our other messages.

Will it be clear from the message which function this is that we can't unwind past? (i.e. from the spans that are shown)

Comment (hyd-dev) [reply compiler/rustc_mir/src/interpret/eval_context.rs:825]: > Will it be clear from the message which function this is that we can't unwind past? (i.e. from the spans that are shown)

I think so, according the Miri test I added: https://github.com/rust-lang/miri/compare/d866c1c52f48bf562720383455b75c257bb1ad92..626638fbfe2fff34648dda29a34d59db498a6e52#diff-523417889d59d37a42fe1b22a2bad6922d9e211bef625d0688f8cb82488cdb0f

Comment (RalfJung) [reply compiler/rustc_mir/src/interpret/eval_context.rs:825]: I can't really tell that from the test, could you show full output?

Comment (hyd-dev) [reply compiler/rustc_mir/src/interpret/eval_context.rs]: Hmm... Will it be unclear if `unwind` is a function pointer?
```rust
#[no_mangle]
extern "C" fn unwind() {
    panic!();
}
fn main() {
    extern "C" {
        fn unwind();
    }
    unsafe { unwind() }
}
```
```
error: Undefined Behavior: unwinding past a frame that does not allow unwinding
 --> output.rs:9:14
  |
9 |     unsafe { unwind() }
  |              ^^^^^^^^ unwinding past a frame that does not allow unwinding
  |
  = help: this indicates a bug in the program: it performed an invalid operation, and caused Undefined Behavior
  = help: see https://doc.rust-lang.org/nightly/reference/behavior-considered-undefined.html for further information
          
  = note: inside `main` at /home/hyd-dev/Playground/output.rs:9:14
  = note: inside `<fn() as std::ops::FnOnce<()>>::call_once - shim(fn())` at $RUST_SRC/library/core/src/ops/function.rs:227:5
  = note: inside `std::sys_common::backtrace::__rust_begin_short_backtrace::<fn(), ()>` at $RUST_SRC/library/std/src/sys_common/backtrace.rs:125:18
  = note: inside closure at $RUST_SRC/library/std/src/rt.rs:49:18
  = note: inside `std::ops::function::impls::<impl std::ops::FnOnce<()> for &dyn std::ops::Fn() -> i32 + std::marker::Sync + std::panic::RefUnwindSafe>::call_once` at $RUST_SRC/library/core/src/ops/function.rs:259:13
  = note: inside `std::panicking::try::do_call::<&dyn std::ops::Fn() -> i32 + std::marker::Sync + std::panic::RefUnwindSafe, i32>` at $RUST_SRC/library/std/src/panicking.rs:401:40
  = note: inside `std::panicking::try::<i32, &dyn std::ops::Fn() -> i32 + std::marker::Sync + std::panic::RefUnwindSafe>` at $RUST_SRC/library/std/src/panicking.rs:365:19
  = note: inside `std::panic::catch_unwind::<&dyn std::ops::Fn() -> i32 + std::marker::Sync + std::panic::RefUnwindSafe, i32>` at $RUST_SRC/library/std/src/panic.rs:433:14
  = note: inside `std::rt::lang_start_internal` at $RUST_SRC/library/std/src/rt.rs:34:21
  = note: inside `std::rt::lang_start::<()>` at $RUST_SRC/library/std/src/rt.rs:48:5

error: aborting due to previous error

Comment (hyd-dev) [reply compiler/rustc_mir/src/interpret/eval_context.rs]: Maybe it's a bit problematic for a function pointer?
```rust
#[no_mangle]
extern "C" fn unwind() {
    panic!();
}

fn get_fn_ptr_from_somewhere_else() -> unsafe extern "C" fn() {
    unwind
}

fn main() {
    let fn_ptr = get_fn_ptr_from_somewhere_else();
    unsafe { fn_ptr() }
}
```
```
error: Undefined Behavior: unwinding past a frame that does not allow unwinding
  --> output.rs:12:14
   |
12 |     unsafe { fn_ptr() }
   |              ^^^^^^^^ unwinding past a frame that does not allow unwinding
   |
   = help: this indicates a bug in the program: it performed an invalid operation, and caused Undefined Behavior
   = help: see https://doc.rust-lang.org/nightly/reference/behavior-considered-undefined.html for further information
           
   = note: inside `main` at output.rs:12:14
   = note: inside `<fn() as std::ops::FnOnce<()>>::call_once - shim(fn())` at $RUST_SRC/library/core/src/ops/function.rs:227:5
   = note: inside `std::sys_common::backtrace::__rust_begin_short_backtrace::<fn(), ()>` at $RUST_SRC/library/std/src/sys_common/backtrace.rs:125:18
   = note: inside closure at $RUST_SRC/library/std/src/rt.rs:49:18
   = note: inside `std::ops::function::impls::<impl std::ops::FnOnce<()> for &dyn std::ops::Fn() -> i32 + std::marker::Sync + std::panic::RefUnwindSafe>::call_once` at $RUST_SRC/library/core/src/ops/function.rs:259:13
   = note: inside `std::panicking::try::do_call::<&dyn std::ops::Fn() -> i32 + std::marker::Sync + std::panic::RefUnwindSafe, i32>` at $RUST_SRC/library/std/src/panicking.rs:401:40
   = note: inside `std::panicking::try::<i32, &dyn std::ops::Fn() -> i32 + std::marker::Sync + std::panic::RefUnwindSafe>` at $RUST_SRC/library/std/src/panicking.rs:365:19
   = note: inside `std::panic::catch_unwind::<&dyn std::ops::Fn() -> i32 + std::marker::Sync + std::panic::RefUnwindSafe, i32>` at $RUST_SRC/library/std/src/panic.rs:433:14
   = note: inside `std::rt::lang_start_internal` at $RUST_SRC/library/std/src/rt.rs:34:21
   = note: inside `std::rt::lang_start::<()>` at $RUST_SRC/library/std/src/rt.rs:48:5

error: aborting due to previous error
```

@rustbot label -S-waiting-on-author S-waiting-on-review

Comment (RalfJung) [reply compiler/rustc_mir/src/interpret/eval_context.rs]: If you move this logic to before the `pop` call, the stack trace should include the called function.

File: compiler/rustc_mir/src/interpret/terminator.rs [hunk 9]
```diff
@@ -230,37 +256,38 @@ impl<'mir, 'tcx: 'mir, M: Machine<'mir, 'tcx>> InterpCx<'mir, 'tcx, M> {
...
+                    caller_abi.name()
+                )
             }
             Ok(())
         };
 
         match instance.def {
             ty::InstanceDef::Intrinsic(..) => {
-                check_abi(self, instance.ty(*self.tcx, self.param_env))?;
+                if M::enforce_abi(self) {
```
Comment (RalfJung) [inline compiler/rustc_mir/src/interpret/terminator.rs:288]: Why did you move the `enforce_abi` check out of the `check_abi` closure?

File: compiler/rustc_mir/src/interpret/terminator.rs [hunk 10]
```diff
@@ -58,12 +69,16 @@ impl<'mir, 'tcx: 'mir, M: Machine<'mir, 'tcx>> InterpCx<'mir, 'tcx, M> {
                 let old_stack = self.frame_idx();
                 let old_loc = self.frame().loc;
                 let func = self.eval_operand(func, None)?;
-                let (fn_val, abi) = match *func.layout.ty.kind() {
+                let (fn_val, abi, can_unwind) = match *func.layout.ty.kind() {
```
Comment (RalfJung) [inline compiler/rustc_mir/src/interpret/terminator.rs:72]: `can_unwind` here is *caller* information; please reflect that in the variable name.

File: compiler/rustc_mir/src/interpret/terminator.rs [hunk 11]
```diff
@@ -281,7 +308,19 @@ impl<'mir, 'tcx: 'mir, M: Machine<'mir, 'tcx>> InterpCx<'mir, 'tcx, M> {
...
+
+                if M::enforce_abi(self) {
+                    check_abi(callee_abi)?;
+                }
+
+                if !matches!(unwind, StackPopUnwind::NotAllowed)
+                    && !self
+                        .fn_can_unwind(self.tcx.codegen_fn_attrs(callee_def_id).flags, callee_abi)
+                {
+                    unwind = StackPopUnwind::NotAllowed;
```
Comment (RalfJung) [inline compiler/rustc_mir/src/interpret/terminator.rs:322]:
```suggestion
                    // The callee cannot unwind.
                    unwind = StackPopUnwind::NotAllowed;
```

File: compiler/rustc_mir/src/interpret/terminator.rs (hunk 9 above)
Comment (hyd-dev) [reply compiler/rustc_mir/src/interpret/terminator.rs:288]: That can avoid calling `get_abi()`.

Comment (RalfJung) [reply compiler/rustc_mir/src/interpret/terminator.rs:288]: Ah, so it's a (small) perf thing? Sure.

Thinking more about this... we still do the "unwinding is UB when caller or callee don't support it" check. So just disabling the ABI check will not be enough to support the testcases this is meant for, will it?

File: compiler/rustc_mir/src/interpret/eval_context.rs (hunk 8 above)
Comment (hyd-dev) [reply compiler/rustc_mir/src/interpret/eval_context.rs]: > If you move this logic to before the `pop` call, the stack trace should include the called function.

:+1: The output looks like this:
```
error: Undefined Behavior: unwinding past a frame that does not allow unwinding
  --> output.rs:2:1
   |
2  | / extern "C" fn unwind() {
3  | |     panic!();
4  | | }
   | |_^ unwinding past a frame that does not allow unwinding
   |
   = help: this indicates a bug in the program: it performed an invalid operation, and caused Undefined Behavior
   = help: see https://doc.rust-lang.org/nightly/reference/behavior-considered-undefined.html for further information
           
   = note: inside `unwind` at output.rs:2:1
note: inside `main` at output.rs:12:14
  --> output.rs:12:14
   |
12 |     unsafe { fn_ptr() }
   |              ^^^^^^^^
   = note: inside `<fn() as std::ops::FnOnce<()>>::call_once - shim(fn())` at $RUST_SRC/library/core/src/ops/function.rs:227:5
   = note: inside `std::sys_common::backtrace::__rust_begin_short_backtrace::<fn(), ()>` at $RUST_SRC/library/std/src/sys_common/backtrace.rs:125:18
   = note: inside closure at $RUST_SRC/library/std/src/rt.rs:49:18
   = note: inside `std::ops::function::impls::<impl std::ops::FnOnce<()> for &dyn std::ops::Fn() -> i32 + std::marker::Sync + std::panic::RefUnwindSafe>::call_once` at $RUST_SRC/library/core/src/ops/function.rs:259:13
   = note: inside `std::panicking::try::do_call::<&dyn std::ops::Fn() -> i32 + std::marker::Sync + std::panic::RefUnwindSafe, i32>` at $RUST_SRC/library/std/src/panicking.rs:401:40
   = note: inside `std::panicking::try::<i32, &dyn std::ops::Fn() -> i32 + std::marker::Sync + std::panic::RefUnwindSafe>` at $RUST_SRC/library/std/src/panicking.rs:365:19
   = note: inside `std::panic::catch_unwind::<&dyn std::ops::Fn() -> i32 + std::marker::Sync + std::panic::RefUnwindSafe, i32>` at $RUST_SRC/library/std/src/panic.rs:433:14
   = note: inside `std::rt::lang_start_internal` at $RUST_SRC/library/std/src/rt.rs:34:21
   = note: inside `std::rt::lang_start::<()>` at $RUST_SRC/library/std/src/rt.rs:48:5

error: aborting due to previous error
```

File: compiler/rustc_mir/src/interpret/terminator.rs (hunk 9 above)
Comment (hyd-dev) [reply compiler/rustc_mir/src/interpret/terminator.rs:288]: > Thinking more about this... we still do the "unwinding is UB when caller or callee don't support it" check. So just disabling the ABI check will not be enough to support the testcases this is meant for, will it?

I think it's enough: https://github.com/rust-lang/miri/commit/cd8f2a57ccadd2d2348d2ae39ed6d577e8cc18d8
Although I have to remove some tests related to function pointers, or that I'm not sure how to *rewrite* without function pointers, since they are always allowed to unwind, in `fn_ptr_codegen_fn_attr_flags()`.

File: compiler/rustc_mir/src/interpret/eval_context.rs [hunk 12]
```diff
@@ -757,13 +757,20 @@ impl<'mir, 'tcx: 'mir, M: Machine<'mir, 'tcx>> InterpCx<'mir, 'tcx, M> {
...
+    /// If `target` is `StackPopUnwind::NotAllowed`, that indicates the function does not allow
+    /// unwinding, and doing so is UB.
+    pub fn unwind_to_block(&mut self, target: StackPopUnwind) -> InterpResult<'tcx> {
         self.frame_mut().loc = match target {
-            Some(block) => Ok(mir::Location { block, statement_index: 0 }),
-            None => Err(self.frame_mut().body.span),
+            StackPopUnwind::Cleanup(block) => Ok(mir::Location { block, statement_index: 0 }),
+            StackPopUnwind::Skip => Err(self.frame_mut().body.span),
+            StackPopUnwind::NotAllowed => {
+                throw_ub_format!("unwinding past a frame that does not allow unwinding")
```
Comment (RalfJung) [inline compiler/rustc_mir/src/interpret/eval_context.rs:770]:
```suggestion
                throw_ub_format!("unwinding past a stack frame that does not allow unwinding")
```

File: compiler/rustc_mir/src/interpret/eval_context.rs [hunk 13]
```diff
@@ -852,11 +851,16 @@ impl<'mir, 'tcx: 'mir, M: Machine<'mir, 'tcx>> InterpCx<'mir, 'tcx, M> {
...
+                StackPopCleanup::Goto { unwind, .. } => unwind,
+                StackPopCleanup::None { .. } => {
+                    panic!("Encountered StackPopCleanup::None when unwinding!")
+                }
+            };
+            self.unwind_to_block(unwind)?;
         } else {
             // Follow the normal return edge.
-            if let Some(ret) = next_block {
+            if let StackPopCleanup::Goto { ret, .. } = return_to_block {
```
Comment (RalfJung) [inline compiler/rustc_mir/src/interpret/eval_context.rs:863]: Please write this as a `match`, so that if we ever add another variant to `StackPopCleanup` we don't forget to adjust this code.

File: compiler/rustc_mir/src/interpret/terminator.rs (hunk 9 above)
Comment (RalfJung) [reply compiler/rustc_mir/src/interpret/terminator.rs:288]: > Although I have to remove some tests related to function pointers, or that I'm not sure how to rewrite without function pointers, since function pointers are always allowed to unwind, in fn_ptr_codegen_fn_attr_flags().

That seems like a bug; a call to an `extern "C" fn()` should not be allowed to unwind.

**Comment by RalfJung** [comment]
LGTM modulo 2 small final nits. :)

**Comment by RalfJung** [comment]
Thanks a lot. :-)
@bors r+

**Comment by klensy** [comment]
Not squashed commits (
There should really be some check to check if there any overwritten changes and suggest merging commits.
